### DISCUSSION TITLE
No completions from plugins

### DISCUSSION THREAD

**Discussion by mustaqimM** [zdharma-continuum/zinit#515] [Q&A] [answered]
### What happened?

Hello, with zinit freshly installed and no changes to `.zshrc`, I'm not getting any completions after running `exec zsh` or starting a new shell. If I install a plugin that has completions like starship by appending: 

```zsh
zinit ice as"command" from"gh-r" \
          atclone"./starship init zsh > init.zsh; ./starship completions zsh > _starship" \
          atpull"%atclone" src"init.zsh"
zinit light starship/starship
```

I get completions while in that shell, but with a new shell, I get no completions.

After installing a plugin with completions I get: 
```zsh
Installed 1 completions. They are stored in the $INSTALLED_COMPS array.
```
And running `zinit completions` show that it's installed.

EDIT. It seems if I manually run `zinit compinit` in a shell, then I get back completions.

### Steps to reproduce

1. Install zinit
2. Install a plugin that has a completion file
3. Reload shell
4. No completion for that plugin when pressing `TAB`

### Operating System & Version

OS: linux-gnu | Vendor: ubuntu | Machine: x86_64 | CPU: x86_64 | Processor: x86_64 | Hardware: x86_64

### Zsh version

zsh 5.9 (x86_64-ubuntu-linux-gnu)

### Terminal emulator

tmux-256color

### If using WSL on Windows, which version of WSL

None

### Additional context

Full `.zshrc`:

```zsh
### Added by Zinit's installer
if [[ ! -f $HOME/.local/share/zinit/zinit.git/zinit.zsh ]]; then
    print -P "%F{33} %F{220}Installing %F{33}ZDHARMA-CONTINUUM%F{220} Initiative Plugin Manager (%F{33}zdharma-continuum/zinit%F{220})…%f"
    command mkdir -p "$HOME/.local/share/zinit" && command chmod g-rwX "$HOME/.local/share/zinit"
    command git clone https://github.com/zdharma-continuum/zinit "$HOME/.local/share/zinit/zinit.git" && \
        print -P "%F{33} %F{34}Installation successful.%f%b" || \
        print -P "%F{160} The clone has failed.%f%b"
fi

source "$HOME/.local/share/zinit/zinit.git/zinit.zsh"
autoload -Uz _zinit
(( ${+_comps} )) && _comps[zinit]=_zinit

# Load a few important annexes, without Turbo
# (this is currently required for annexes)
zinit light-mode for \
    zdharma-continuum/zinit-annex-as-monitor \
    zdharma-continuum/zinit-annex-bin-gem-node \
    zdharma-continuum/zinit-annex-patch-dl \
    zdharma-continuum/zinit-annex-rust

### End of Zinit's installer chunk

# Load starship theme
# line 1: `starship` binary as command, from github release
# line 2: starship setup at clone(create init.zsh, completion)
# line 3: pull behavior same as clone, source init.zsh
zinit ice as"command" from"gh-r" \
          atclone"./starship init zsh > init.zsh; ./starship completions zsh > _starship" \
          atpull"%atclone" src"init.zsh"
zinit light starship/starship
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by vladdoster** [comment, marked as answer]
@mustaqimM,

This is a configuration issue and not a bug. Furthermore, it isn't related to zinit either, but how to use Zsh. 

As you have observed, completions work when you run `zinit compinit`. If we look at the code for the `compinit` command, it runs the following code:

<img width="1272" alt="Screenshot 2023-04-30 at 17 46 28" src="https://user-images.githubusercontent.com/10052309/235379506-fb4226a1-0a54-4673-92c7-39fb5fee4f1c.png">

It calls the `.zinit-compinit` function, so let us look at what it does. To do this, you can use the `functions` (it outputs a functions source code) command and run the following:

```zsh
functions -x4 .zinit-compinit
```

I've highlighted the code essential to the issue at hand.

<img width="1415" alt="Screenshot 2023-04-30 at 17 53 40" src="https://user-images.githubusercontent.com/10052309/235379700-cf96298a-2b4e-423f-8fe8-929bfe6ee40f.png">

In the [Zsh documentation for completion](https://zsh.sourceforge.io/Doc/Release/Completion-System.html#Initialization), here is what `compinit` does.

> This section describes the use of compinit to initialize completion for the current session when called directly; if you have run compinstall it will be called automatically from your .zshrc.

The issue is you don't initialize the completion system in your `.zshrc`. See the [`completions` section in the README](https://github.com/zdharma-continuum/zinit#completions-2) for this.

I'd recommend using turbo mode. All you need to do is add the following snippet to the end of your zshrc.

```zsh
zi for \
    atload"zicompinit; zicdreplay" \
    blockf \
    lucid \
    wait \
  @zsh-users/zsh-completions
```

**Reply by mustaqimM** [reply]
@vladdoster Thank you for this. Sorry, you're right. In the minimal config I didn't have the snippet above and in my full config I had `atinit"ZINIT[COMPINIT_OPTS]=-C; zicompinit; zicdreplay"`. Removing `ZINIT[COMPINIT_OPTS]=-C;` brings back completion at the cost of loading speed. It's been a while since I touched the config so I'll have to see what's up:

```zsh
zi wait lucid light-mode for \
  atclone'(){local f;cd -q →*;for f (*~*.zwc){zcompile -Uz -- ${f}};}' \
  atpull'%atclone' compile'.*fast*~*.zwc' atinit"zpcompinit; zpcdreplay" \
    zdharma-continuum/fast-syntax-highlighting \
  atinit"
    bindkey '^[OA' history-substring-search-up
    bindkey '^P'   history-substring-search-up
    bindkey '^[OB' history-substring-search-down
    bindkey '^N'   history-substring-search-down" \
      zsh-users/zsh-history-substring-search \
  compile'{src/*.zsh,src/strategies/*}' atload'!_zsh_autosuggest_start' \
    zsh-users/zsh-autosuggestions \
  atpull'zinit creinstall -q ~/.zsh/completions' blockf \
    zsh-users/zsh-completions
```

EDIT. The slow loading speed was due to not setting `skip_global_compinit=1` on Ubuntu

**Reply by vladdoster** [reply]
@mustaqimM,

Awesome. You can find my [config here](https://github.com/vladdoster/dotfiles/blob/master/zsh/.config/zsh/rc.d/04-zinit.zsh) or the official [zdharma-continuum/zinit-configs](https://github.com/zdharma-continuum/zinit-configs) for inspiration.
